core/src/main/scala/org/apache/spark/deploy/security/HiveDelegationTokenProvider.scala

```diff
@@ -85,7 +85,10 @@ private[spark] class HiveDelegationTokenProvider
       val principal = conf.getTrimmed(principalKey, "")
       require(principal.nonEmpty, s"Hive principal $principalKey undefined")
       val metastoreUri = conf.getTrimmed("hive.metastore.uris", "")
-      require(metastoreUri.nonEmpty, "Hive metastore uri undefined")
+      if (metastoreUri.isEmpty) {
```

Comment:
How is the code getting past the check in `delegationTokensRequired`? It's basically checking for the same thing.

Reply:
require() will throws IllegalArgumentException and exits the JVM here. Letting `delegationTokensRequired` return when metastoreUri is undefined (using JDBC to connect DB directly) only finishes this method (not set token in Credentials).

Reply:
Same question as @vanzin , `delegationTokensRequired` should already check whether `hive.metastore.uris` is empty or not, so it will not obtain the DT if this `hive.metastore.uris` is not configured.

```scala
  override def delegationTokensRequired(
      sparkConf: SparkConf,
      hadoopConf: Configuration): Boolean = {
    // Delegation tokens are needed only when:
    // - trying to connect to a secure metastore
    // - either deploying in cluster mode without a keytab, or impersonating another user
    //
    // Other modes (such as client with or without keytab, or cluster mode with keytab) do not need
    // a delegation token, since there's a valid kerberos TGT for the right user available to the
    // driver, which is the only process that connects to the HMS.
    val deployMode = sparkConf.get("spark.submit.deployMode", "client")
    UserGroupInformation.isSecurityEnabled &&
      hiveConf(hadoopConf).getTrimmed("hive.metastore.uris", "").nonEmpty &&
      (SparkHadoopUtil.get.isProxyUser(UserGroupInformation.getCurrentUser()) ||
        (deployMode == "cluster" && !sparkConf.contains(KEYTAB)))
  }
```

Reply:
Yes, I know. If the metastore is undefined, it no needs to obtain the DT. Am I right?

Reply:
Before getting the DT, we will check if `delegationTokensRequired` returns true or false, if it is false, then we will not get DT. Here since "hive.metastore.uris" is not configured, then it should return false.

Reply:
@jerryshao can I use #21343 to replace #20784 ? in #20784 , it didn't check `delegationTokensRequired`

Reply:
Oh, my fault, `delegationTokensRequired` has been checked in `SparkSQLCLIDriver.scala`.